### DISCUSSION TITLE
Improve CI and PR process

### DISCUSSION THREAD

**Discussion by Mirko-von-Leipzig** [0xMiden/node#425] [General]
Some ideas, not all are good :)

---------------------------

Restrict workflows to only run the latest commit in a PR. Code example [here](https://github.com/eqlabs/pathfinder/blob/59cd89be7b233d226e7ab044840125f871b4b769/.github/workflows/ci.yml#L13-L15). Essentially this cancels any in-progress workflows for the PR whenever it is updated. This prevents notifications for failed workflows from old commits for example. In addition to saving CI time as workflows are queued.

---------------------------

Switch the default branch to `next`. This will become the branch PRs are opened against by default. Downside is that cloning also defaults to this branch, meaning users will clone the wrong branch.

---------------------------

Cache the CI runs, this will drastically speed up CI times. Tests currently compile in ~2 minutes or roughly half the CI total time. Caching even just the latest `next` branch would improve every PR and commit. This also applies to lint and doc workflows and is probably more impactful there.

---------------------------

Consider a [codeowners](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) file.

**Comment by bobbinth** [comment]
> Restrict workflows to only run the latest commit in a PR. Code example [here](https://github.com/eqlabs/pathfinder/blob/59cd89be7b233d226e7ab044840125f871b4b769/.github/workflows/ci.yml#L13-L15). Essentially this cancels any in-progress workflows for the PR whenever it is updated. This prevents notifications for failed workflows from old commits for example. In addition to saving CI time as workflows are queued.

This would be great! I don't see any downsides to doing this.

> Switch the default branch to `next`. This will become the branch PRs are opened against by default. Downside is that cloning also defaults to this branch, meaning users will clone the wrong branch.

We considered this a couple of times in the past and the main reason we kept it the way it is now is that we want the default branch be consistent with what is on crates.io (i.e., a kind of "latest stable").

> Cache the CI runs, this will drastically speed up CI times. Tests currently compile in ~2 minutes or roughly half the CI total time. Caching even just the latest `next` branch would improve every PR and commit. This also applies to lint and doc workflows and is probably more impactful there.

This would be awesome and we actually have an issue for something similar in [miden-base](https://github.com/0xPolygonMiden/miden-base/issues/568).

> Consider a [codeowners](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) file.

I haven't used this in the past - but it seems like a good idea.

**Comment by Mirko-von-Leipzig** [comment]
The caching and workflow improvements have been implemented. `next` branch left as is.
